fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/utility/handler/ElectionHandlerTest.java

```diff
@@ -287,22 +338,114 @@ public void testHandleElectionEnd()
+
+    assertEquals(Hash.hash(voteIds), election.computeRegisteredVotesHash());
+  }
+
+  @Test
+  public void castVoteOnlyKeepsLastVote()
+      throws UnknownElectionException, UnknownRollCallException, UnknownLaoException,
+          DataHandlingException, NoRollCallException {
+    handleElectionSetup(OPEN_BALLOT_ELECTION, LAO_CHANNEL);
+    handleElectionKey(OPEN_BALLOT_ELECTION, ELECTION_KEY);
+    handleElectionOpen(OPEN_BALLOT_ELECTION);
+
+    // Handle the votes of two different senders. One sender sends two votes, and the first should
+    // be discarded
+    handleCastVote(VOTE1, SENDER_KEY, OPENED_AT);
+    handleCastVote(VOTE2, ATTENDEE_KEY, OPENED_AT + 1);
+    handleCastVote(VOTE3, ATTENDEE_KEY, OPENED_AT + 2);
+
+    // The expected hash is made on the sorted vote ids (check that vote2 was discarded)
+    String[] voteIds = Stream.of(VOTE1, VOTE3).map(Vote::getId).sorted().toArray(String[]::new);
     Election election = electionRepo.getElectionByChannel(OPEN_BALLOT_ELECTION.getChannel());
 
     assertEquals(Hash.hash(voteIds), election.computeRegisteredVotesHash());
   }
 
+  @Test
+  public void castVoteDiscardsStaleVote()
+      throws UnknownElectionException, UnknownRollCallException, UnknownLaoException,
+          DataHandlingException, NoRollCallException {
+    handleElectionSetup(OPEN_BALLOT_ELECTION, LAO_CHANNEL);
+    handleElectionKey(OPEN_BALLOT_ELECTION, ELECTION_KEY);
+    handleElectionOpen(OPEN_BALLOT_ELECTION);
+
+    // Handle the votes of two different senders. One sender sends two votes, and the second should
+    // be discarded as it has an older creation date
+    handleCastVote(VOTE1, SENDER_KEY, OPENED_AT);
+    handleCastVote(VOTE2, ATTENDEE_KEY, OPENED_AT + 2);
+    handleCastVote(VOTE3, ATTENDEE_KEY, OPENED_AT + 1);
+
+    // The expected hash is made on the sorted vote ids (check that vote3 was discarded)
+    String[] voteIds = Stream.of(VOTE1, VOTE2).map(Vote::getId).sorted().toArray(String[]::new);
+    Election election = electionRepo.getElectionByChannel(OPEN_BALLOT_ELECTION.getChannel());
+
+    assertEquals(Hash.hash(voteIds), election.computeRegisteredVotesHash());
+  }
+
+  @Test
+  public void castVoteFailsOnPreviousDataNull()
+      throws UnknownElectionException, UnknownRollCallException, UnknownLaoException,
+          DataHandlingException, NoRollCallException {
+    handleElectionSetup(OPEN_BALLOT_ELECTION, LAO_CHANNEL);
+    handleElectionKey(OPEN_BALLOT_ELECTION, ELECTION_KEY);
+    handleElectionOpen(OPEN_BALLOT_ELECTION);
+
+    MessageGeneral nullData = new MessageGeneral(SENDER_KEY, null, gson);
+    setMessageRepoReturns(nullData);
+
+    handleCastVote(VOTE1, SENDER_KEY, OPENED_AT);
+    assertThrows(
+        IllegalStateException.class, () -> handleCastVote(VOTE1, SENDER_KEY, OPENED_AT + 1));
+  }
+
+  @Test
+  public void castVoteFailsOnPreviousDataInvalid()
+      throws UnknownElectionException, UnknownRollCallException, UnknownLaoException,
```

Comment:
I think this test function deserves a comment explaining what you test. It is not immediately clear from the code and the function name IMO